File: lib/models/email.dart
class Email {
  final String sender;
  final String subject;
  final String preview;
  final String body;
  final String time;
  bool isRead;
  bool isStarred;
  String category;
  List<String> labels;
  bool hasAttachments;
  String? id;

  Email({
    required this.sender,
    required this.subject,
    required this.preview,
    required this.body,
    required this.time,
    this.isRead = false,
    this.isStarred = false,
    this.category = 'primary',
    List<String>? labels,
    this.hasAttachments = false,
    this.id,
  }) : labels = labels ?? [];

  factory Email.fromMap(Map<String, dynamic> map) {
    return Email(
      sender: map['sender'] ?? '',
      subject: map['subject'] ?? '',
      preview: map['preview'] ?? '',
      body: map['body'] ?? '',
      time: map['time'] ?? '',
      isRead: map['isRead'] ?? false,
      isStarred: map['isStarred'] ?? false,
      category: map['category'] ?? 'primary',
      labels: List<String>.from(map['labels'] ?? []),
      hasAttachments: map['attachments'] == true || map['hasAttachments'] == true,
      id: map['id'],
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'sender': sender,
      'subject': subject,
      'preview': preview,
      'body': body,
      'time': time,
      'isRead': isRead,
      'isStarred': isStarred,
      'category': category,
      'labels': labels,
      'attachments': hasAttachments,
      'id': id,
    };
  }
} 

File: models/Email.js
const mongoose = require('mongoose');

const emailSchema = new mongoose.Schema({
  from: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true }, // User._id of sender
  to: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },   // User._id of primary recipient
  cc: [{ type: mongoose.Schema.Types.ObjectId, ref: 'User', default: [] }],    // Array of User._id
  bcc: [{ type: mongoose.Schema.Types.ObjectId, ref: 'User', default: [] }],   // Array of User._id
  subject: { type: String, required: true },
  body: { type: String, required: true },
  time: { type: String, required: true },
  isRead: { type: Boolean, default: false },
  isStarred: { type: Boolean, default: false },
  category: { type: String, default: 'inbox', enum: ['inbox', 'sent', 'draft', 'trash', 'archive', 'spam'] },
  labels: [{ type: String }],
  hasAttachments: { type: Boolean, default: false },
  attachmentUrls: [{ type: String }],
  inReplyTo: { type: mongoose.Schema.Types.ObjectId, ref: 'Email', default: '' }, // Email._id
  forwardedFrom: { type: mongoose.Schema.Types.ObjectId, ref: 'Email', default: '' }, // Email._id
});

// Add indexes for frequently queried fields
emailSchema.index({ from: 1, to: 1, category: 1, labels: 1 });

module.exports = mongoose.model('Email', emailSchema);